[GTK][WPE] Error while creating the vblank monitor

Scony opened the pull request:
#### 6036e80ef9789aac67d02a939f9f8ee34a39e4f8
<pre>
[GTK][WPE] Error while creating the vblank monitor
<a href="https://bugs.webkit.org/show_bug.cgi?id=276952">https://bugs.webkit.org/show_bug.cgi?id=276952</a>

Reviewed by Carlos Garcia Campos.

This PR changes DisplayVBlankMonitorDRM::create() in a way that drm node with crtc
is being seeked instead of just looking for crtc based on custom drm node.

* Source/WebKit/UIProcess/glib/DisplayVBlankMonitorDRM.cpp:
(WebKit::findDrmNodeWithCrtc):
(WebKit::DisplayVBlankMonitorDRM::create):

Canonical link: <a href="https://commits.webkit.org/281290@main">https://commits.webkit.org/281290@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b27fd4eeb03f3b7e251c8894ea17f7f9064d138

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59408 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38751 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11929 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63322 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/9918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46405 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10082 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48233 "Passed tests") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/9918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61438 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36195 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51420 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29058 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32903 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8855 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54854 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8953 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65055 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3336 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8867 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55576 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3347 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51415 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55680 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/document/load-events (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2770 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8868 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34567 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/35650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36736 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35395 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayVBlankMonitorDRM.cpp:
```diff
@@ -182,6 +182,45 @@ static std::optional<std::pair<uint32_t, uint32_t>> findCrtc(int fd)
+static std::optional<std::pair<DrmNode, CrtcInfo>> findDrmNodeWithCrtc(PlatformMonitor* monitor = nullptr)
+#else
+static std::optional<std::pair<DrmNode, CrtcInfo>> findDrmNodeWithCrtc()
+#endif
```
Maybe for this it would be clearer to use a helper Struct instead of a pairs

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayVBlankMonitorDRM.cpp:
```diff
@@ -182,6 +182,45 @@ static std::optional<std::pair<uint32_t, uint32_t>> findCrtc(int fd)
+static std::optional<std::pair<DrmNode, CrtcInfo>> findDrmNodeWithCrtc()
+#endif
+{
+    constexpr int maxDrmDevices = 256;
```
This is not the maximum number of devices, but the maximum nodes (3x64 rounded to 256), but here we want the maximum number of devices including 3 possible nodes, so it should be 64.

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayVBlankMonitorDRM.cpp:
```diff
@@ -182,6 +182,45 @@ static std::optional<std::pair<uint32_t, uint32_t>> findCrtc(int fd)
+#endif
+{
+    constexpr int maxDrmDevices = 256;
+    drmDevicePtr devices[maxDrmDevices] = { 0 };
```
I don't think we need to initialize this.

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayVBlankMonitorDRM.cpp:
```diff
@@ -182,6 +182,45 @@ static std::optional<std::pair<uint32_t, uint32_t>> findCrtc(int fd)
+{
+    constexpr int maxDrmDevices = 256;
+    drmDevicePtr devices[maxDrmDevices] = { 0 };
+    const int devicesNum = drmGetDevices(devices, maxDrmDevices);
```
Use `drmGetDevices2` and pass 0 as flags since we don't need PCI revision.

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayVBlankMonitorDRM.cpp:
```diff
@@ -182,6 +182,45 @@ static std::optional<std::pair<uint32_t, uint32_t>> findCrtc(int fd)
+    if (devicesNum <= 0)
+        return { };
+    for (int i = 0; i < devicesNum; i++) {
+        if (devices[i]->available_nodes & (1 << DRM_NODE_PRIMARY)) {
```
I would continue if it doesn't have a primary node and we save an indentation level for the rest of the for body.

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayVBlankMonitorDRM.cpp:
```diff
@@ -182,6 +182,45 @@ static std::optional<std::pair<uint32_t, uint32_t>> findCrtc(int fd)
+#endif
+            if (crtcInfo) {
+                drmFreeDevices(devices, devicesNum);
+                return std::make_pair(fd, *crtcInfo);
```
`WTFMove(fd)` we don't want to duplicate the descriptor.

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayVBlankMonitorDRM.cpp:
```diff
@@ -216,49 +255,67 @@ std::unique_ptr<DisplayVBlankMonitor> DisplayVBlankMonitorDRM::create(PlatformDi
+        if (!fd) {
+            RELEASE_LOG_FAULT(DisplayLink, "Could not create a vblank monitor for display %u: failed to open %s", displayID, filename.utf8().data());
+            return nullptr;
+        }
```
For the case of DRM, I think it would be simpler if we create a drmNodeWithCrtcInfo, initialized with the fd here, and adding later the crtc info, that way we can handle it the same way and avoid more ifdefs.

carlosgcampos commented on Source/WebKit/UIProcess/glib/DisplayVBlankMonitorDRM.cpp:
```diff
@@ -216,63 +256,62 @@ std::unique_ptr<DisplayVBlankMonitor> DisplayVBlankMonitorDRM::create(PlatformDi
-            crtcInfo = { wpe_monitor_drm_get_crtc_index(WPE_MONITOR_DRM(monitor)), wpe_monitor_get_refresh_rate(monitor) };
-        else
+        if (!WPE_IS_MONITOR_DRM(monitor))
 #endif
```
So, at this point `drmNodeWithCrtcInfo` can only be initialized for DRM platform, so we can avoid this ifdef and DRM monitor check and simply do:

```
if (!drmNodeWithCrtcInfo)
    drmNodeWithCrtcInfo = findDrmNodeWithCrtc(monitor);
```

Scony committed:
[GTK][WPE] Error while creating the vblank monitor
https://bugs.webkit.org/show_bug.cgi?id=276952

Reviewed by Carlos Garcia Campos.

This PR changes DisplayVBlankMonitorDRM::create() in a way that drm node with crtc
is being seeked instead of just looking for crtc based on custom drm node.

* Source/WebKit/UIProcess/glib/DisplayVBlankMonitorDRM.cpp:
(WebKit::findDrmNodeWithCrtc):
(WebKit::DisplayVBlankMonitorDRM::create):

Canonical link: https://commits.webkit.org/281290@main
